# Limpieza del dataset de bienes raíces

Este es un conjunto de datos (dataset) reales que fue descargado usando técnicas de web scraping. El archivo contiene registros de **Fotocasa**, el cual es uno de los sitios más populares de bienes raíces en España. Contiene miles de datos de casas reales publicadas en la web www.fotocasa.com.

El dataset fue descargado hace algunos años y en ningún caso se obtuvo beneficio económico de ello.

Tu objetivo es extraer tanta información como sea posible con el conocimiento que tienes hasta ahora de ciencia de datos.

Agrega observaciones, conclusiones o notas acerca de los datos obtenidos en cada ejercicio.

In [115]:
import pandas as pd

#### Ejercicio 00

Lee el dataset data/real_estate.csv e intenta visualizar la tabla (★☆☆)

In [116]:
# Este archivo CSV contiene puntos y comas en lugar de comas como separadores
df = pd.read_csv('assets/real_estate.csv', sep=';')
df

Unnamed: 0  id_realEstates  isNew                    realEstate_name  \
0               1       153771986  False            ferrari 57 inmobiliaria   
1               2       153867863  False  tecnocasa fuenlabrada ferrocarril   
2               3       153430440  False                 look find boadilla   
3               4       152776331  False  tecnocasa fuenlabrada ferrocarril   
4               5       153180188  False            ferrari 57 inmobiliaria   
...           ...             ...    ...                                ...   
15330       15331       153901377  False                infocasa consulting   
15331       15332       150394373  False                inmobiliaria pulpon   
15332       15333       153901397  False             tecnocasa torrelodones   
15333       15334       152607440  False                inmobiliaria pulpon   
15334       15335       153901356  False                infocasa consulting   

       phone_realEstate                                       url_inmueble  \
0           912177526.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
1           916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
2           916350408.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
3           916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
4           912177526.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
...                 ...                                                ...   
15330       911360461.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15331       912788039.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15332       912780348.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15333       912788039.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15334       911360461.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   

       rooms  bathrooms  surface   price  ... level4Id level5Id level6Id  \
0        3.0        2.0    103.0  195000  ...        0        0        0   
1        3.0        1.0      NaN   89000  ...        0        0        0   
2        2.0        2.0     99.0  390000  ...        0        0        0   
3        3.0        1.0     86.0   89000  ...        0        0        0   
4        2.0        2.0    106.0  172000  ...        0        0        0   
...      ...        ...      ...     ...  ...      ...      ...      ...   
15330    2.0        1.0     96.0  259470  ...        0        0        0   
15331    3.0        1.0    150.0  165000  ...        0        0        0   
15332    4.0        2.0    175.0  495000  ...        0        0        0   
15333    3.0        2.0    101.0  195000  ...        0        0        0   
15334    3.0        2.0    152.0  765000  ...        0        0        0   

      level7Id level8Id accuracy          latitude          longitude zipCode  \
0            0        0        0  40,2948276786438  -3,44402412135624     NaN   
1            0        0        1          40,28674           -3,79351     NaN   
2            0        0        0  40,4115646786438  -3,90662252135624     NaN   
3            0        0        0  40,2853785786438  -3,79508142135624     NaN   
4            0        0        0  40,2998774864376  -3,45226301356237     NaN   
...        ...      ...      ...               ...                ...     ...   
15330        0        0        0          40,45416           -3,70286     NaN   
15331        0        0        0          40,36652           -3,48951     NaN   
15332        0        0        0          40,57444           -3,92124     NaN   
15333        0        0        0          40,36967           -3,48105     NaN   
15334        0        0        0          40,45773           -3,69068     NaN   

      customZone  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
15330        NaN  
15331        NaN  
15332        NaN  
15333        NaN  
15334        NaN  

[15335

## Trabajando con un DataFrame

#### Ejercicio 01

¿Cuál es la casa más cara del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada. Para visualizar el resultado utiliza un f-string. Por ejemplo:

```py
f'La casa más cara se encuentra en la dirección: {address} y su precio es {price} €'
```

In [117]:
casa_cara = df['price'].idxmax()
fila_casa_cara = df.loc[casa_cara]
calle = fila_casa_cara['address']
precio = fila_casa_cara['price']
results = {'casa_cara': calle, 'precio':int(precio)}
results

{'casa_cara': 'El Escorial', 'precio': 8500000}

#### Ejercicio 02

¿Cuál es la casa más barata del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada utilizando f-string

In [118]:
casa_barata = df['price'].idxmin()
fila_casa_barata = df.loc[casa_barata]
calle2 = fila_casa_barata['address']
precio2 = fila_casa_barata['price']
results = {'casa_barata' : calle2, 'precio': precio2}
results

{'casa_barata': 'Parla', 'precio': np.int64(0)}

#### Ejercicio 03

¿Cuál es la casa más grande del dataset? (★☆☆)

Imprime la dirección y el área de las casas seleccionadas utilizando f-string

In [119]:
casa_superficie_grande = df['surface'].idxmax()
fila_casa_superficie_grande = df.loc[casa_superficie_grande]
direccion = fila_casa_superficie_grande['address']
superficie = fila_casa_superficie_grande['surface']
results = {'casa_grande_direccion' : direccion, 'superficie': superficie}
results

{'casa_grande_direccion': 'Sevilla la Nueva',
 'superficie': np.float64(249000.0)}

#### Ejercicio 04

¿Cuál es la casa más pequeña del dataset? (★☆☆)

In [120]:
casa_superficie_pequeña = df['surface'].idxmin()
fila_casa_superficie_pequeña = df.loc[casa_superficie_pequeña]
direccion2 = fila_casa_superficie_pequeña['address']
superficie2 = fila_casa_superficie_pequeña['surface']
results = {'casa_pequeña_direccion' : direccion2, 'superficie': superficie}
results

{'casa_pequeña_direccion': 'Calle Amparo,  Madrid Capital',
 'superficie': np.float64(249000.0)}

#### Ejercicio 05.

¿El dataset contiene valores no admitidos (NAs)? (★☆☆)

- Muestra el nombre de las filas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.
- También muestra el nombre de las columnas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.

In [121]:
filas_nas = df.isnull().any(axis=1)
columnas_nas = df.isnull().any()
filas_nas
columnas_nas


Unnamed: 0          False
id_realEstates      False
isNew               False
realEstate_name      True
phone_realEstate     True
url_inmueble        False
rooms                True
bathrooms            True
surface              True
price               False
date                False
description          True
address             False
country             False
level1              False
level2              False
level3              False
level4               True
level5              False
level6               True
level7               True
level8               True
upperLevel          False
countryId           False
level1Id            False
level2Id            False
level3Id            False
level4Id            False
level5Id            False
level6Id            False
level7Id            False
level8Id            False
accuracy            False
latitude            False
longitude           False
zipCode              True
customZone           True
dtype: bool

#### Ejercicio 06.

Elimina las columnas que contengan NAs del dataset, si aplica (★★☆)

Muestra las dimensiones del DataFrame original y del DataFrame después de las eliminaciones.

In [122]:
data_frame = df.dropna(axis=1)
results = {'df_original' : {'filas': df.shape[0], 'columnas' : df.shape[1]}, 
           'df_modificado': {'filas': data_frame.shape[0], 'columnas' : data_frame.shape[1]}}
results

{'df_original': {'filas': 15335, 'columnas': 37},
 'df_modificado': {'filas': 15335, 'columnas': 25}}

#### Ejercicio 07

¿Cuantas poblaciones (columna level5) contiene el dataset? (★☆☆)

- Muestra una lista con los nombres de las poblaciones
- Muestra el total de las mismas

In [123]:
poblaciones = df['level5'].unique()
total_poblaciones = len(poblaciones)
results = {'total_poblaciones' : total_poblaciones, 'poblaciones' : poblaciones}
results


{'total_poblaciones': 168,
 'poblaciones': array(['Arganda del Rey', 'Fuenlabrada', 'Boadilla del Monte',
        'Las Rozas de Madrid', ' Madrid Capital', 'Villaviciosa de Odón',
        'Pinto', 'Valdemoro', 'Navalcarnero', 'Pozuelo de Alarcón',
        'Torrejón de Ardoz', 'Navalagamella', 'San Sebastián de los Reyes',
        'Rivas-vaciamadrid', 'Alpedrete', 'Móstoles',
        'San Fernando de Henares', 'Coslada', 'Becerril de la Sierra',
        'Alcalá de Henares', 'Chinchón', 'Parla', 'Alcorcón',
        'El Escorial', 'Leganés', 'Pedrezuela', 'Majadahonda',
        'Villanueva de la Cañada', 'Villanueva del Pardillo',
        'Torrelodones', 'Moralzarzal', 'Mejorada del Campo', 'Aranjuez',
        'Corpa', 'Getafe', 'Velilla de San Antonio', 'Sevilla la Nueva',
        'San Martín de la Vega', 'Villalbilla', 'Collado Villalba',
        'Alcobendas', 'El Molar (Madrid)', 'Moraleja de Enmedio', 'Algete',
        'Campo Real', 'Torrejón de la Calzada', 'Colmenar Viejo',
        

#### Ejercicio 08

¿Cuál es la media de precios en la población (columna level5) de "Arroyomolinos (Madrid)"? (★★☆)

In [124]:
poblacion_arroyomolinos = df['level5'] == 'Arroyomolinos (Madrid)'
df_arroyomolinos = df.loc[poblacion_arroyomolinos]
media_precios = df_arroyomolinos['price'].mean()
results = round(media_precios, 2)
results


np.float64(294541.6)

#### Ejercicio 09.

¿Los precios promedios de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

- Muestra ambos promedios
- Escribe en una celda markdown una conclusión sobre ellos

In [131]:
valdemorillo = df['level5'] == 'Valdemorillo'
galapagar = df['level5'] == 'Galapagar'
df_valdemorillo = df[valdemorillo]
df_galapagar = df[galapagar]

promedio_v = df_valdemorillo['price'].mean()
promedio_g = df_galapagar['price'].mean()

diferencia = promedio_v - promedio_g

results = {'mean_valdemorillo' : promedio_v, 'mean_galapagar' : promedio_g, 'difference_mean' : diferencia}
results

{'mean_valdemorillo': np.float64(363860.2931034483),
 'mean_galapagar': np.float64(360063.20238095237),
 'difference_mean': np.float64(3797.090722495923)}

A pesar de ser poblaciones distintas ambas comparten un promedio de precios muy parecidos, con una diferencia de 3797.09€ lo que indica que comparten ciertas características similares.

#### Ejercicio 10

¿Los promedios de precio por metro cuadrado (precio/m2) de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

> Pista: Crea una nueva columna llamada `pps` (*price per square* o precio por metro cuadrado) y luego analiza los valores.

- Muestra ambos promedios de precio por metro cuadrado
- Escribe en una celda markdown una conclusión sobre ellos

In [132]:
df['pps'] = df['price'] / df['surface']

promedio_v_pps = df_valdemorillo['pps'].mean()
promedio_g_pps = df_galapagar['pps'].mean()

diferencia_porcentual = round(((promedio_g_pps - promedio_v_pps) / promedio_v_pps) * 100, 2)
diferencia_porcentual

results = {'m2_valdemorillo' : promedio_v_pps, 'm2_galapagar' : promedio_g_pps, 'difference_per' : diferencia_porcentual}
results

{'m2_valdemorillo': np.float64(1317.9502109024986),
 'm2_galapagar': np.float64(1606.3240303094024),
 'difference_per': np.float64(21.88)}

Ambas poblaciones tienen un precio promedio similar, sin embargo hay una gran diferencia de valor en el m2 por superficie, del 21,88%.
Esto es un indicativo que en las casas de Valdemorillo ofrece una mejor relación superficie-precio, el m2 por superficie se está vendiendo notablemente más barato que las casas ubicadas en Galapagar.

#### Ejercicio 11

¿Cuántas agencia de bienes raíces contiene el dataset? (★★☆)

- Muestra el valor obtenido.

In [133]:
count_real_estate = df['realEstate_name'].unique()
count_real_estate.shape

(1822,)

In [134]:
count_real_estate

array(['ferrari 57 inmobiliaria', 'tecnocasa fuenlabrada ferrocarril',
       'look find boadilla', ..., 'redpiso villaverde bajo san cristobal',
       'gestion comercial', 'nieves'], shape=(1822,), dtype=object)

#### Ejercicio 12

¿Cuál es la población (columna level5) que contiene la mayor cantidad de casas?(★★☆)

- Muestra la población y el número de casas.

In [138]:
poblaciones = df['level5'].value_counts()
poblacion_max = poblaciones.index[0]
casas_max = poblaciones.iloc[0]
results = {'poblacion_max' : poblacion_max, 'casas': casas_max}
results


{'poblacion_max': ' Madrid Capital', 'casas': np.int64(6643)}

---

## Trabajando con un subconjunto del DataFrame

#### Ejercicio 13

Ahora vamos a trabajar con el "cinturón sur" de Madrid.

Haz un subconjunto del DataFrame original que contenga las siguientes poblaciones (columna level5): "Fuenlabrada", "Leganés", "Getafe", "Alcorcón" (★★☆)

> Pista: Filtra el DataFrame original usando la columna `level5` y la función `isin`.

In [ ]:
masc_lvl5 = df['level5'].isin(['Fuenlabrada', 'Leganés', 'Getafe', 'Alcorcón'])
df_sub = df[masc_lvl5] 
df_sub

Unnamed: 0  id_realEstates  isNew  \
1               2       153867863  False   
3               4       152776331  False   
85             86       153152077  False   
94             95       153995577  False   
109           110       153586414  False   
...           ...             ...    ...   
15275       15276       153903887  False   
15291       15292       151697757  False   
15305       15306       153902389  False   
15322       15323       153871864  False   
15325       15326       153901467  False   

                                 realEstate_name  phone_realEstate  \
1              tecnocasa fuenlabrada ferrocarril       916358736.0   
3              tecnocasa fuenlabrada ferrocarril       916358736.0   
85                       sinergical inmobiliaria               NaN   
94                               viviendas365com       911226014.0   
109              area uno asesores inmobiliarios       912664081.0   
...                                          ...               ...   
15275  aliseda servicios de gestion inmobiliaria       911368198.0   
15291                                    unipiso       912788631.0   
15305                             jadein ferrero       914871639.0   
15322                          gestion comercial       911220662.0   
15325                              montehogar 68       911790675.0   

                                            url_inmueble  rooms  bathrooms  \
1      https://www.fotocasa.es/es/comprar/vivienda/ma...    3.0        1.0   
3      https://www.fotocasa.es/es/comprar/vivienda/ma...    3.0        1.0   
85     https://www.fotocasa.es/es/comprar/vivienda/le...    1.0        1.0   
94     https://www.fotocasa.es/es/comprar/vivienda/le...    3.0        2.0   
109    https://www.fotocasa.es/es/comprar/vivienda/ma...    3.0        3.0   
...                                                  ...    ...        ...   
15275  https://www.fotocasa.es/es/comprar/vivienda/al...    3.0        1.0   
15291  https://www.fotocasa.es/es/comprar/vivienda/al...    3.0        2.0   
15305  https://www.fotocasa.es/es/comprar/vivienda/ma...    3.0        2.0   
15322  https://www.fotocasa.es/es/comprar/vivienda/ma...    3.0        1.0   
15325  https://www.fotocasa.es/es/comprar/vivienda/ma...    2.0        2.0   

       surface   price  ... level5Id level6Id level7Id level8Id accuracy  \
1          NaN   89000  ...        0        0        0        0        1   
3         86.0   89000  ...        0        0        0        0        0   
85        50.0  107000  ...        0        0        0        0        1   
94       120.0  320000  ...        0        0        0        0        0   
109      142.0  425000  ...        0        0        0        0        0   
...        ...     ...  ...      ...      ...      ...      ...      ...   
15275     78.0  138000  ...        0        0        0        0        1   
15291    110.0  279000  ...        0        0        0        0        0   
15305     85.0  170000  ...        0        0        0        0        0   
15322     91.0  112000  ...        0        0        0        0        0   
15325     99.0  215000  ...        0        0        0        0        1   

               latitude          longitude zipCode customZone          pps  
1              40,28674           -3,79351     NaN        NaN          NaN  
3      40,2853785786438  -3,79508142135624     NaN        NaN  1034.883721  
85             40,35059           -3,82693     NaN        NaN  2140.000000  
94             40,31933           -3,77574     NaN        NaN  2666.666667  
109          40,3313411         -3,8313868     NaN        NaN  2992.957746  
...                 ...                ...     ...        ...          ...  
15275          40,31381           -3,83733     NaN        NaN  1769.230769  
15291        40,3259051           -3,76318     NaN        NaN  2536.363636  
15305        40,2882193         -3,8098617     NaN        NaN  2000.000000  
15322          40

#### Ejercicio 14

Calcula la media y la varianza de muestra para las siguientes variables: precio, habitaciones, superficie y baños (★★★)

> Debes usar el subset obtenido en la pregunta 13

- Crea y visualiza un diccionario con todos los valores

In [ ]:
variables = ['price', 'rooms', 'surface', 'bathrooms']
resultados = round(df_sub[variables].agg(['var', 'mean']), 2)
diccionario = (resultados.to_dict('dict'))
diccionario

{'price': {'var': 14921367508.05, 'mean': 223094.48},
 'rooms': {'var': 0.72, 'mean': 3.02},
 'surface': {'var': 4263.05, 'mean': 111.75},
 'bathrooms': {'var': 0.57, 'mean': 1.63}}

#### Ejercicio 15

¿Cuál es la casa más cara de cada población del cinturón sur de Madríd? (★★☆)

> Debes usar el subset obtenido en la pregunta 13

- Genera un DataFrame con esta información
- Muestra tanto la dirección como el precio de la casa seleccionada de cada población.
- Genera conclusiones en una celda markdown

In [ ]:
df_fuenlabrada = df_sub[df_sub['level5'] == 'Fuenlabrada']
df_leganes = df_sub[df_sub['level5'] == 'Leganés']
df_getafe = df_sub[df_sub['level5'] == 'Getafe']
df_alcorcon = df_sub[df_sub['level5'] == 'Alcorcón']

df_ordenado = df_sub.sort_values('price', ascending = False)
df_casas_caras = df_ordenado.drop_duplicates('level5')
df_sur = df_casas_caras[['level5', 'address', 'price']]
df_sur



level5                                address    price
2881        Getafe                                 Getafe  1050000
5585      Alcorcón                               Alcorcón   950000
10412      Leganés           Avenida Reina Sofía, Leganés   650000
11561  Fuenlabrada  Calle de Paulo Freire, 5, Fuenlabrada   490000

>***OBSERVACIONES***

> Estos valores altos representan los puntos atipicos, los chalets de lujo de las poblaciones.

#### Ejercicio 16

¿Qué puedes decir acerca del precio por metro cuadrado (precio/m2) entre los municipios de 'Getafe' y 'Alcorcón'? (★★☆)

> Debes usar el subset obtenido en la pregunta 13
>
> Pista: Crea una nueva columna llamada `pps` (price per square en inglés) y luego analiza los valores

In [ ]:
average_getafe  = round(df_getafe['pps'].mean(), 2)
average_alcorcon = round(df_alcorcon['pps'].mean() ,2)

diferencia_porcentual = round(((average_alcorcon - average_getafe) / average_getafe) * 100, 2)

results = {'promedio_m2_getafe': average_getafe, 'promedio_m2_alcorcon' : average_alcorcon, 'diferencia_per' : diferencia_porcentual}
results

{'promedio_m2_getafe': np.float64(2066.31),
 'promedio_m2_alcorcon': np.float64(2239.3),
 'diferencia_per': np.float64(8.37)}

>***OBSERVACIONES***

> "La diferencia porcentual del precio promedio por metro cuadrado entre Getafe y Alcorcón es relativamente baja (8,37€). Esto se debe a que ambas poblaciones pertenecen al cinturón sur de Madrid y comparten un contexto inmobiliario similar. Por lo tanto, el mercado las valora de manera muy pareja, reflejando precios por metro cuadrado muy competitivos y homogéneos."

## Conclusiones

#### Ejercicio 17

Escribe aquí tus conclusiones acerca de este proyecto

>***OBSERVACIONES***

> El proyecto ayuda al estudiante de forma escalada a familiarisarse y entender como filtrar y acceder a los elementos de un dataframe.
> Este proyecto ha servido para adquirir una comprensión fundamental del análisis exploratorio de datos (EDA) y la manipulación de Pandas en Python. 